docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,287 @@
+
+1. if `missingReplicas` > 0 or `missingReplicas` = -1 and this subset is not unschedulable, choose this subset temporarily 
+   
+   (1). update `missingReplicas` -= 1 and put this pod into `creatingPods` map and try to update subsetStatus. \
+   (2). If update to WorkloadSpread conflicts with other updates, get the WorkloadSpread again and go back to step (1). \
+   (3). if update to WorkloadSpread successfully, inject the rules of this subset into the Pod.
+   
+2. if `missingReplicas` = 0, go to the next subset
+   
+3. if there is no available subset left, just let the Pod go
+
+##### Pod deletion 
+Also, when a Pod that belongs to a subset is being deleted or evicted, webhook will put it into the `deletingPods` map in subsetStatus and update `missingReplicas` += 1.
+
+##### Update Pod Status
+When a Pod that belongs to a subset changes status phase to 'succeed' or 'failed', which means Pod has been terminated lifecycle, webhook will clean it from `creatingPods` map and update `missingReplicas` += 1.
+
+#### Reschedule strategy
+Reschedule strategy will delete unscheduled Pods that still in pending status. Some subsets have no sufficient resource can lead to some Pods unscheduable.
+WorkloadSpread has multiple subset, so the unschedulable pods should be rescheduled to other subsets.
+
+Controller will mark the subset containing unscheduable pods to unscheduable status and Webhook cannot inject Pod into this subset by check subset's status.
+And then controller cleans up all unscheduable Pods to trigger workload creating new replicas, and webhook will skip unscheduable subset and chose suitable subset to inject. 
+
+The unscheduled subset can be kept for 10 minutes and then should be recovered schedulable to schedule Pod again by controller.
+
+#### Deletion priority
+
+We have three types for subset's Pod deletion-cost
+1. the number of active Pods in this subset <= maxReplicas, deletion-cost = 100. indicating the priority of Pods \
+   in this subset is higher than other Pods in workload.
+2. subset's maxReplicas is nil, deletion-cost = 0. indicating the priority of Pods in this subset is lower than \
+   other subsets that have a not nil maxReplicas.
+3. the number of active Pods in this subset > maxReplicas, two class: (a) deletion-cost = -100, (b) deletion-cost = +100. \
+   indicating we prefer deleting the Pod have -100 deletion-cost in this subset in order to control the instance of subset \
+   meeting up the desired maxReplicas number.
+
+CloneSet has supported deletion-cost feature in the recent versions.
+
+Since Kubernetes 1.21, there is a new annotation definition on Pod:
+
+```yaml
+    controller.kubernetes.io/pod-deletion-cost: -100 / 100 / 0
+```
+
+(In 1.21, users need to enable `PodDeletionCost` feature-gate, and since 1.22 it will be enabled by default)
+
+### Alternative Considered
+
+#### UnitedDeployment
+
+Both UnitedDeployment and WorkloadSpread allow the workload to be distributed in different subsets. 
```

Comment:
i think the key difference between UnitedDeployment and WorkloadSpread is that WorkloadSpread requires NO workload api change, one can use vanilla CloneSet even Deployment while attaching extra topology constraints. The UnitedDeployment can also be improved to understand the subset preferences